nixos/test-driver: use an argument instead of the out env-var

marijanp opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Bug  #153766, read [comment](https://github.com/NixOS/nixpkgs/issues/153766#issuecomment-1007295433)

###### Things done
Instead of reading the `out` environment variable in the test-driver's Machine::copy_from_vm and Machine::screenshot functions, a output-directory is used.
The output-directory is a mandatory command-line argument that needs to be passed to the test-driver.
When makeTest invokes the test-driver the `out` environment variable is passed as the argument.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

marijanp committed:
nixos/test-driver: use an argument instead of the out env-var

Synthetica9 commented:
Oh drat, I just implemented my own fix as well... If we don't end up going with that, I also cleaned up the screenshot logic, we should probably incorporate that here as well.

marijanp commented:
@ofborg test nat.firewall networking.scripted.link installer.simpleProvided installer.separateBootFat networking.scripted.virtual printing keymap.azerty keymap.dvorak-programmer boot-stage1 installer.swraid keymap.neo nfs4.simple i3wm udisks2 networking.networkd.loopback containers-ip ecryptfs login installer.simpleLabels php.httpd zfs.installer predictable-interface-names.unpredictable predictable-interface-names.unpredictableNetworkd mutableUsers

marijanp commented:
> Oh drat, I just implemented my own fix as well... If we don't end up going with that, I also cleaned up the screenshot logic, we should probably incorporate that here as well.

We could split it up into two PRs each for the respective change?

Synthetica9 commented on nixos/lib/test-driver/test_driver/__init__.py:
```diff
@@ -63,6 +75,13 @@ def main() -> None:
         help="vlans to span by the driver",
     )
+    arg_parser.add_argument(
+        "output_directory",
```
Having this as a positional argument breaks the current interface of allowing `nix build -f . nixosTests.foo && ./result/bin/nixos-test-driver`, would prefer a `--output-directory` flag with default value `$PWD`

marijanp replied:
Do you mean `nix build -f . nixosTests.foo.driver && ./result/bin/nixos-test-driver`?  in that case one just needs to run `nix build -f . nixosTests.foo.driver && ./result/bin/nixos-test-driver . `
I made it positional, because I wanted to make use of the WriteableDir action which checks whether the out-dir exists. Since actions are not executed for default values I removed it. But I will make it optional as you suggest and check whether the out-dir is writeable right after parsing the arguments

Synthetica9 commented on nixos/lib/test-driver/test_driver/__init__.py:
```diff
@@ -33,6 +33,18 @@ def __call__(self, parser, namespace, values, option_string=None):  # type: igno
         setattr(namespace, self.dest, values)
 
 
+class WriteableDir(argparse.Action):  # type: ignore
+    def __call__(self, parser, namespace, values, option_string=None):  # type: ignore
+        if not values.is_dir():
+            raise argparse.ArgumentTypeError("{0} is not a directory".format(values))
+        if os.access(values, os.W_OK):
+            setattr(namespace, self.dest, values)
+        else:
+            raise argparse.ArgumentTypeError(
+                "{0} is not a writeable directory".format(values)
+            )
```
If we're gonna introduce this mechanism, can we also use it for `$TEMPDIR`?

marijanp replied:
I did that, but I needed to redefine itsince it should not raise a argparse specific exception

marijanp committed:
make output_directory an optional parameter

marijanp committed:
introduce writeable_dir argparse type

marijanp committed:
add tmp_dir access check

marijanp committed:
factor out tmp-dir checks

marijanp commented:
@ofborg test nat.firewall networking.scripted.link installer.simpleProvided installer.separateBootFat networking.scripted.virtual printing keymap.azerty keymap.dvorak-programmer boot-stage1 installer.swraid keymap.neo nfs4.simple i3wm udisks2 networking.networkd.loopback containers-ip ecryptfs login installer.simpleLabels php.httpd zfs.installer predictable-interface-names.unpredictable predictable-interface-names.unpredictableNetworkd mutableUsers

marijanp commented:
@ofborg test pantheon boot.biosCdrom boot.uefiUsb networking.networkd.static latestKernel.login networking.networkd.dhcpSimple networking.networkd.sit env switchTest hibernate predictable-interface-names.predictable gnome networking.networkd.bridge networking.networkd.privacy plasma5 systemd-networkd-ipv6-prefix-delegation lightdm installer.simple installer.btrfsSimple installer.btrfsSubvols sway networking.networkd.link networking.scripted.privacy

marijanp commented:
@ofborg test nat.firewall-conntrack networking.scripted.sit networking.networkd.virtual networking.scripted.dhcpSimple firewall firefox networking.scripted.macvlan proxy firefox-esr sddm.default boot.uefiCdrom networking.scripted.static networking.scripted.routes installer.lvm networking.networkd.vlan containers-imperative boot.biosUsb xfce networking.networkd.dhcpOneIf installer.simpleUefiSystemdBoot keymap.qwertz gnome-xorg shadow networking.networkd.macvlan networking.networkd.routes

marijanp commented:
@ofborg test nfs3.simple php.fpm fontconfig-default-fonts networking.scripted.bond nano keymap.dvorak installer.luksroot networking.scripted.bridge docker simple chromium networking.scripted.dhcpOneIf networking.networkd.bond php.pcre nat.standalone openssh networking.scripted.vlan ipv6 installer.separateBoot misc predictable-interface-names.predictableNetworkd networking.scripted.loopback keymap.colemak installer.btrfsSubvolDefault

Synthetica9 commented:
This broke test output, now screenshots and such are not copied to $out. #156910 fixes this.
